### ISSUE TITLE
export cache picks random version among bib(la)tex formats

### ISSUE THREAD

**Issue by mmccourt** [ZotPlus/zotero-better-bibtex#205]
I've just updated to Zotero 4.0.26.4 and the Better Bib(La)Tex addon 0.10.25.

Now when I export a collection using either the Better Bibtex or the Better BibLatex format, the "year" tag is missing in the xml.

The result is that references in the pdf documents have "????" where the year should be e.g.
[Granovetter(????b)]
and
[Granovetter(????b)] Granovetter,Mark S., ????b, The strength of weak ties 1360–1380.

(I am using MikTex 2.9 editor.)

Working old .bib file entry:
@article{Granovetter:1973,
title = {{The strength of weak ties}},
url = {http://www.jstor.org.ezp.essec.fr/stable/2776392},
urldate = {2014-12-04},
journal = {{American journal of sociology}},
author = {Granovetter, Mark S.},
year = {1973},
pages = {1360--1380}
}

Broken new .bib file:
@article{Granovetter:1973,
title = {The strength of weak ties},
url = {http://www.jstor.org.ezp.essec.fr/stable/2776392},
timestamp = {2014-12-04 05:47:42},
journaltitle = {American journal of sociology},
shortjournal = {Am. J. Sociol.},
author = {Granovetter, Mark S.},
urldate = {2014-12-04},
date = {1973},
pages = {1360--1380}
}

I could probably manually edit the .bib file to add the "year" tag, but it would be great if there was a more permanent solution.

**Comment by retorquere** [comment]
Are you *sure* you're using the right exporter? The "year" tag is only
output using the better bibtex translator. Not saying this bug can't be
real (most likely culprit is the caching system), just want to make sure.

**Comment by retorquere** [comment]
I've published 0.10.29 which has extensive debug facilities. Can you export one such entry in which you saw the problem (even if you don't now) by right-clicking it and selecting "Report ZotPlus error"?

**Comment by retorquere** [comment]
The entries you show are from different translators, so I'm afraid there's a cache mixup happening somewhere.

**Comment by retorquere** [comment]
0.10.30 is out, with even more logging. If you could please go into preferences-> advanced, enable "extended zotplus debug logging", and then just export with the Better BibTeX translator. Once you spot the error, select "report zotplus error" from the gear men.

**Comment by mmccourt** [comment]
hi,

thanks for the quick response. I have switched on logging, exported my
collection, and submitted error report id http://is.gd/nUMrNL

However now I think I know what the problem could be, or at least the
behavior required to fix it.

The papers that I added yesterday to the collection (via the Chrome plugin)
have the error. Today (after uninstalling reinstalling Zotero etc) I tried
adding the same papers again using the Chrome plugin - this time they are
good!

So the problem seems to have "resolved itself", the only downside being I
have to re-add a number of papers that I added recently.

Thanks again for your help,
Maurice




On 29 May 2015 at 11:34, retorquere <notifications@github.com> wrote:

> 0.10.30 is out, with even more logging. If you could please go into
> preferences-> advanced, enable "extended zotplus debug logging", and then
> just export with the Better BibTeX translator. Once you spot the error,
> select "report zotplus error" from the gear men.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ZotPlus/zotero-better-bibtex/issues/205#issuecomment-106758480>
> .
>



-- 
m:+33 (0) 787 18.36.54
e:maurice.mccourt@gmail.com
---------------------------------------------

**Comment by mmccourt** [comment]
Actually I may have been too hasty - it seems all the papers (100+) in my
existing collections are not being exported correctly (missing the ""year"
tag).

Just the ones I downloaded today are working.

So I would have to re-download _all_ my existing papers - clearly a major
task.

Is there a better way to approach this?


thanks,
Maurice




On 29 May 2015 at 11:59, Maurice McCourt <maurice.mccourt@gmail.com> wrote:

> hi,
>
> thanks for the quick response. I have switched on logging, exported my
> collection, and submitted error report id http://is.gd/nUMrNL
>
> However now I think I know what the problem could be, or at least the
> behavior required to fix it.
>
> The papers that I added yesterday to the collection (via the Chrome
> plugin) have the error. Today (after uninstalling reinstalling Zotero etc)
> I tried adding the same papers again using the Chrome plugin - this time
> they are good!
>
> So the problem seems to have "resolved itself", the only downside being I
> have to re-add a number of papers that I added recently.
>
> Thanks again for your help,
> Maurice
>
>
>
>
> On 29 May 2015 at 11:34, retorquere <notifications@github.com> wrote:
>
>> 0.10.30 is out, with even more logging. If you could please go into
>> preferences-> advanced, enable "extended zotplus debug logging", and then
>> just export with the Better BibTeX translator. Once you spot the error,
>> select "report zotplus error" from the gear men.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/ZotPlus/zotero-better-bibtex/issues/205#issuecomment-106758480>
>> .
>>
>
>
>
> --
> m:+33 (0) 787 18.36.54
> e:maurice.mccourt@gmail.com
> ---------------------------------------------
>



-- 
m:+33 (0) 787 18.36.54
e:maurice.mccourt@gmail.com
---------------------------------------------

**Comment by retorquere** [comment]
I stil suspect the caching. Can you right-click one of those failing entries and select "report ZotPlus error"? The fact that a re-download helped is in fact encouraging, as it would mean the cache fetcher is OK, but the cache builder is or was making errors. The report will help me in the diagnosis.

You don't have to download papers anew; the data is all still there, it's just the cache that is corrupt. You have two ways to fix this; for both you'll need to go into `about:config`

1. Set `extensions.zotero.translators.better-bibtex.cacheReset` to `true` and restart firefox. This will trigger a one-time cache rebuild. If the error is in the cache building (which I hope to learn from the error report you could generate), this should fix it.
2. If not, you can set `extensions.zotero.translators.better-bibtex.caching` to `false`. This will disable one of the two caches in BBT (the one which I think is the culprit).

Once I track down the error, I'll cut a release in the 0.11 range; this will automatically trigger a one-time cache rebuild, so everyone would get to benefit.

Thank you for your patience and your help.

**Comment by mmccourt** [comment]
hi,

I submitted an error report for one of the failing entries
http://is.gd/h9MCCg

The cachereset (otion 1 in your email) seems to have resolved the problem.

thanks,
Maurice


On 29 May 2015 at 12:37, retorquere <notifications@github.com> wrote:

> I stil suspect the caching. Can you right-click one of those failing
> entries and select "report ZotPlus error"? The fact that a re-download
> helped is in fact encouraging, as it would mean the cache fetcher is OK,
> but the cache builder is or was making errors. The report will help me in
> the diagnosis.
>
> You don't have to download papers anew; the data is all still there, it's
> just the cache that is corrupt. You have two ways to fix this; for both
> you'll need to go into about:config
>
>    1. Set extensions.zotero.translators.better-bibtex.cacheReset to true
>    and restart firefox. This will trigger a one-time cache rebuild. If the
>    error is in the cache building (which I hope to learn from the error report
>    you could generate), this should fix it.
>    2. If not, you can set
>    extensions.zotero.translators.better-bibtex.caching to false. This
>    will disable one of the two caches in BBT (the one which I think is the
>    culprit).
>
> Once I track down the error, I'll cut a release in the 0.11 range; this
> will automatically trigger a one-time cache rebuild, so everyone would get
> to benefit.
>
> Thank you for your patience and your help.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ZotPlus/zotero-better-bibtex/issues/205#issuecomment-106769152>
> .
>



-- 
m:+33 (0) 787 18.36.54
e:maurice.mccourt@gmail.com
---------------------------------------------

**Comment by retorquere** [comment]
OK, that looks good so far. Could you do the following for a single entry:

1. Export using Better BibTeX
2. Export using Better BibLaTeX, using the same export options
3. Export using Better BibTeX again, using the same export options
4. Right-click, and report errors.

It might be this was a bug that has been around for a while, and a single cache rebuild would fix it... yet it's a little suspicious I get two reports of the same in such a short time. I'll cut a 0.11 this afternoon just to be on the safe side, so everyone's cache will be reset, but I wouldn't want to call this one solved just yet.

**Comment by retorquere** [comment]
Wait: http://is.gd/h9MCCg was generated *before* the cache reset, but exported the "year" field when `Better BibTeX` was chosen? That's can't be right. I see two cached entries, both for `Better BibLaTeX`, none for `Better BibTeX`.

**Comment by mmccourt** [comment]
hi

http://is.gd/h9MCCg was generated for an item in the exported collection
*before* the cache reset, and did not export the "year" before the reset.
(Note that the error report is not for the Collection which was exported -
see below)

After the cahce reset the year was exported correctly.

Does that make sense?


I also did the export sequence as you requested. I submitted an error
report this time for the Collection that was exported  http://is.gd/9nN8Fi
(the previous error reports were for an item in the exported Collection)



On 29 May 2015 at 14:09, retorquere <notifications@github.com> wrote:

> Wait: http://is.gd/h9MCCg was generated *before* the cache reset, but
> exported the "year" field when Better BibTeX was chosen? That's can't be
> right. I see two cached entries, both for Better BibLaTeX, none for Better
> BibTeX.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ZotPlus/zotero-better-bibtex/issues/205#issuecomment-106784666>
> .
>



-- 
m:+33 (0) 787 18.36.54
e:maurice.mccourt@gmail.com
---------------------------------------------

**Comment by retorquere** [comment]
It doesn't make sense to me yet. The export at http://is.gd/h9MCCg only has BBLT entries cached. Those entries are not supposed to have the `year` field, so they look OK in principle, but it looks like they got used for the BBT export... which would mean the cache fetcher is broken, not the cache builder. 

The 2nd export doesn't make sense to me either. I only see BBT entries there, no BBLT entries. Looking through the logs, I only see activity for the Better BibTeX and the LaTeX Citation translators, and none for the Better BibLaTeX translators. This explains the cache having only BBT entries, but is inexplicable if you did in fact export BBLT in step 2.

**Comment by mmccourt** [comment]
OK - I followed the sequence again, this time for 1 single item (not a
Collection):
- export Better Bibtex
- export Better Biblatex
- export Better Bibtex

always using the same export settings (ie no options - nothing ticked)

The report ID is http://is.gd/UoTL0f

Hope this helps....

On 29 May 2015 at 20:06, retorquere <notifications@github.com> wrote:

> It doesn't make sense to me yet. The export at http://is.gd/h9MCCg only
> has BBLT entries cached. Those entries are not supposed to have the year
> field, so they look OK in principle, but it looks like they got used for
> the BBT export... which would mean the cache fetcher is broken, not the
> cache builder.
>
> The 2nd export doesn't make sense to me either. I only see BBT entries
> there, no BBLT entries. Looking through the logs, I only see activity for
> the Better BibTeX and the LaTeX Citation translators, and none for the
> Better BibLaTeX translators. This explains the cache having only BBT
> entries, but is inexplicable if you did in fact export BBLT in step 2.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ZotPlus/zotero-better-bibtex/issues/205#issuecomment-106892279>
> .
>



-- 
m:+33 (0) 787 18.36.54
e:maurice.mccourt@gmail.com
---------------------------------------------

**Comment by retorquere** [comment]
Got it. Wow, that was nasty. Fix tonight.

**Comment by retorquere** [comment]
I've changed the issue title to reflect the actual problem... you guys were right about what happened, sorry for doubting you, and thanks for reporting what would have bitten a lot of people otherwise.

Fortunately, this problem was isolated in the cache, it was not data-destructive (but then, BBT updates actual data only in one very isolated spot -- to store keys in the extra field, until Zotero gets around to offering plugins a place to put their stuff and have it sync). In any case, cache stores were OK (which means I won't have to put out 0.12 to trigger a cache rebuild), but cache fetches were essentially random... AAMOF it was an implementation artifact of the Firefox javascript engine that you guys even got the entries for the same reference.

The problem was cause by a misunderstanding on my part of one of the javascript libraries I use for the cache. I've updated the docs for that library and offered a pull request to the author, now merged, that makes my previous way of searching the cache through that library supported; in the interim, I've added a shim to my code to implement the same.

All tests are green, and a release build is running on Travis at https://travis-ci.org/ZotPlus/zotero-better-bibtex/builds/64736405; once that goes green it will publish 0.11.1 automatically.

Thank you both for your patience and tenacity.
